{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CORD-19 Literature Mining\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib as plt\n",
    "from sentence_transformers import SentenceTransformer\n",
    "from nltk import word_tokenize\n",
    "from nltk.corpus import stopwords\n",
    "from transformers import AutoTokenizer, AutoModelForQuestionAnswering\n",
    "import torch\n",
    "import csv\n",
    "import os\n",
    "import json\n",
    "import random\n",
    "from sentence_transformers import util\n",
    "from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Versions\n",
      "============================\n",
      "pandas:  1.3.0\n",
      "numpy:  1.19.2\n",
      "matplotlib:  3.4.2\n"
     ]
    }
   ],
   "source": [
    "print(\"Versions\")\n",
    "print(\"============================\")\n",
    "print(\"pandas: \",pd.__version__)\n",
    "print(\"numpy: \",np.__version__)\n",
    "print(\"matplotlib: \",plt.__version__)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data preprocessing\n",
    "### Filter & clean metacsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Gerard\\anaconda3\\envs\\issenv\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3418: DtypeWarning: Columns (15) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cord_uid</th>\n",
       "      <th>source_x</th>\n",
       "      <th>title</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>authors</th>\n",
       "      <th>journal</th>\n",
       "      <th>pdf_json_files</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ug7v899j</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Clinical features of culture-proven Mycoplasma...</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>OBJECTIVE: This retrospective chart review des...</td>\n",
       "      <td>2001-07-04</td>\n",
       "      <td>Madani, Tariq A; Al-Ghamdi, Aisha A</td>\n",
       "      <td>BMC Infect Dis</td>\n",
       "      <td>document_parses/pdf_json/d1aafb70c066a2068b027...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>02tnwd4m</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Nitric oxide: a pro-inflammatory mediator in l...</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>Inflammatory diseases of the respiratory tract...</td>\n",
       "      <td>2000-08-15</td>\n",
       "      <td>Vliet, Albert van der; Eiserich, Jason P; Cros...</td>\n",
       "      <td>Respir Res</td>\n",
       "      <td>document_parses/pdf_json/6b0567729c2143a66d737...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ejv2xln0</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Surfactant protein-D and pulmonary host defense</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>Surfactant protein-D (SP-D) participates in th...</td>\n",
       "      <td>2000-08-25</td>\n",
       "      <td>Crouch, Erika C</td>\n",
       "      <td>Respir Res</td>\n",
       "      <td>document_parses/pdf_json/06ced00a5fc04215949aa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2b73a28n</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Role of endothelin-1 in lung disease</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>Endothelin-1 (ET-1) is a 21 amino acid peptide...</td>\n",
       "      <td>2001-02-22</td>\n",
       "      <td>Fagan, Karen A; McMurtry, Ivan F; Rodman, David M</td>\n",
       "      <td>Respir Res</td>\n",
       "      <td>document_parses/pdf_json/348055649b6b8cf2b9a37...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9785vg6d</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Gene expression in epithelial cells in respons...</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>Respiratory syncytial virus (RSV) and pneumoni...</td>\n",
       "      <td>2001-05-11</td>\n",
       "      <td>Domachowske, Joseph B; Bonville, Cynthia A; Ro...</td>\n",
       "      <td>Respir Res</td>\n",
       "      <td>document_parses/pdf_json/5f48792a5fa08bed9f560...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cord_uid source_x                                              title  \\\n",
       "0  ug7v899j      PMC  Clinical features of culture-proven Mycoplasma...   \n",
       "1  02tnwd4m      PMC  Nitric oxide: a pro-inflammatory mediator in l...   \n",
       "2  ejv2xln0      PMC    Surfactant protein-D and pulmonary host defense   \n",
       "3  2b73a28n      PMC               Role of endothelin-1 in lung disease   \n",
       "4  9785vg6d      PMC  Gene expression in epithelial cells in respons...   \n",
       "\n",
       "  license                                           abstract publish_time  \\\n",
       "0   no-cc  OBJECTIVE: This retrospective chart review des...   2001-07-04   \n",
       "1   no-cc  Inflammatory diseases of the respiratory tract...   2000-08-15   \n",
       "2   no-cc  Surfactant protein-D (SP-D) participates in th...   2000-08-25   \n",
       "3   no-cc  Endothelin-1 (ET-1) is a 21 amino acid peptide...   2001-02-22   \n",
       "4   no-cc  Respiratory syncytial virus (RSV) and pneumoni...   2001-05-11   \n",
       "\n",
       "                                             authors         journal  \\\n",
       "0                Madani, Tariq A; Al-Ghamdi, Aisha A  BMC Infect Dis   \n",
       "1  Vliet, Albert van der; Eiserich, Jason P; Cros...      Respir Res   \n",
       "2                                    Crouch, Erika C      Respir Res   \n",
       "3  Fagan, Karen A; McMurtry, Ivan F; Rodman, David M      Respir Res   \n",
       "4  Domachowske, Joseph B; Bonville, Cynthia A; Ro...      Respir Res   \n",
       "\n",
       "                                      pdf_json_files  \n",
       "0  document_parses/pdf_json/d1aafb70c066a2068b027...  \n",
       "1  document_parses/pdf_json/6b0567729c2143a66d737...  \n",
       "2  document_parses/pdf_json/06ced00a5fc04215949aa...  \n",
       "3  document_parses/pdf_json/348055649b6b8cf2b9a37...  \n",
       "4  document_parses/pdf_json/5f48792a5fa08bed9f560...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import metacsv\n",
    "date_cols = ['publish_time']\n",
    "data = pd.read_csv(r'.\\Cord19\\archive/metadata.csv',\n",
    "                  usecols=['cord_uid','source_x','title','license','publish_time',\n",
    "                           'abstract','authors','journal','pdf_json_files'], parse_dates=date_cols)\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of articles: 311274\n",
      "Dropped 66738 articles with missing abstract\n",
      "Dropped 151478 articles with missing pdf json\n",
      "Dropped 9525 articles with missing journal name\n",
      "Dropped 80 articles with missing authors\n",
      "Dropped 0 articles with missing publish time\n",
      "Dropped 0 articles with missing title"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Gerard\\anaconda3\\envs\\issenv\\lib\\site-packages\\pandas\\util\\_decorators.py:311: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  return func(*args, **kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Dropped 166 articles with duplicate abstract and uid\n",
      "Data rows: 83287\n"
     ]
    }
   ],
   "source": [
    "# Filter by date\n",
    "start_date = \"2021-1-1\"\n",
    "end_date = \"2022-1-31\"\n",
    "\n",
    "after_start_date = data[\"publish_time\"] >= start_date\n",
    "before_end_date = data[\"publish_time\"] <= end_date\n",
    "between_two_dates = after_start_date & before_end_date\n",
    "data1 = data.loc[between_two_dates]\n",
    "print(f'Number of articles: {(len(data1))}')\n",
    "\n",
    "# Data cleaning: remove missing abstracts, missing pdf json, duplicate abstract and uid\n",
    "len1 = len(data1)\n",
    "data1.dropna(axis=0,how='any',subset=['abstract'],inplace=True)\n",
    "print(f'Dropped {len1-len(data1)} articles with missing abstract')\n",
    "\n",
    "len1 = len(data1)\n",
    "data1.dropna(axis=0,how='any',subset=['pdf_json_files'],inplace=True)\n",
    "print(f'Dropped {len1-len(data1)} articles with missing pdf json')\n",
    "\n",
    "len1 = len(data1)\n",
    "data1.dropna(axis=0,how='any',subset=['journal'],inplace=True)\n",
    "print(f'Dropped {len1-len(data1)} articles with missing journal name')\n",
    "\n",
    "len1 = len(data1)\n",
    "data1.dropna(axis=0,how='any',subset=['authors'],inplace=True)\n",
    "print(f'Dropped {len1-len(data1)} articles with missing authors')\n",
    "\n",
    "len1 = len(data1)\n",
    "data1.dropna(axis=0,how='any',subset=['publish_time'],inplace=True)\n",
    "print(f'Dropped {len1-len(data1)} articles with missing publish time')\n",
    "\n",
    "len1 = len(data1)\n",
    "data1.dropna(axis=0,how='any',subset=['title'],inplace=True)\n",
    "print(f'Dropped {len1-len(data1)} articles with missing title')\n",
    "\n",
    "len1 = len(data1)\n",
    "data1.drop_duplicates(subset=['abstract'],keep='first',inplace=True)\n",
    "data1.drop_duplicates(subset=['cord_uid'],keep='first',inplace=True)\n",
    "print(f'Dropped {len1-len(data1)} articles with duplicate abstract and uid')\n",
    "\n",
    "print(f'Data rows: {len(data1)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Gerard\\anaconda3\\envs\\issenv\\lib\\site-packages\\pandas\\core\\frame.py:3607: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._set_item(key, value)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cord_uid</th>\n",
       "      <th>source_x</th>\n",
       "      <th>title</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>authors</th>\n",
       "      <th>journal</th>\n",
       "      <th>pdf_json_files</th>\n",
       "      <th>abstract_tokens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14597</th>\n",
       "      <td>aoiambb8</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Etiology of Severe Acute Respiratory Infection...</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>in april 2017, surveillance detected a surge i...</td>\n",
       "      <td>2021-01-27</td>\n",
       "      <td>Rahaman, Md R.; Alroy, Karen A.; Van Beneden, ...</td>\n",
       "      <td>Emerg Infect Dis</td>\n",
       "      <td>document_parses/pdf_json/3798d3bcb09df78296832...</td>\n",
       "      <td>[april, 2017, surveillance, detected, surge, s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14601</th>\n",
       "      <td>embv1el9</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Accuracy of a battery-powered portable capnome...</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>carbon dioxide measurement is useful for confi...</td>\n",
       "      <td>2021-01-01</td>\n",
       "      <td>Hirakawa, Eiji; Ibara, Satoshi</td>\n",
       "      <td>J Clin Monit Comput</td>\n",
       "      <td>document_parses/pdf_json/2b4690b639d799f33bbf7...</td>\n",
       "      <td>[carbon, dioxide, measurement, useful, confirm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14602</th>\n",
       "      <td>8dcgzyst</td>\n",
       "      <td>PMC</td>\n",
       "      <td>The influence of diabetes on postoperative com...</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>background: diabetes mellitus has been commonl...</td>\n",
       "      <td>2021-01-01</td>\n",
       "      <td>Tan, D. J. H.; Yaow, C. Y. L.; Mok, H. T.; Ng,...</td>\n",
       "      <td>Tech Coloproctol</td>\n",
       "      <td>document_parses/pdf_json/83a841d74d4f6bf8bffb8...</td>\n",
       "      <td>[background, diabetes, mellitus, commonly, ass...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14603</th>\n",
       "      <td>sgtvecj8</td>\n",
       "      <td>PMC</td>\n",
       "      <td>In vivo delivery of a multiepitope peptide and...</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>objectives: a potent hiv vaccine should overco...</td>\n",
       "      <td>2021-01-01</td>\n",
       "      <td>Davoodi, Saba; Bolhassani, Azam; Namazi, Fatemeh</td>\n",
       "      <td>Biotechnol Lett</td>\n",
       "      <td>document_parses/pdf_json/7dd8097ba18590caa4310...</td>\n",
       "      <td>[objectives, potent, hiv, vaccine, overcome, l...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14604</th>\n",
       "      <td>3vtq6d8f</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Relationship Between Compassion Fatigue in Nur...</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>this research was conducted in order to examin...</td>\n",
       "      <td>2021-01-02</td>\n",
       "      <td>Aslan, Hakime; Erci, Behice; Pekince, Hatice</td>\n",
       "      <td>J Relig Health</td>\n",
       "      <td>document_parses/pdf_json/9bd542f156f108ef170a0...</td>\n",
       "      <td>[research, conducted, order, examine, correlat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       cord_uid source_x                                              title  \\\n",
       "14597  aoiambb8      PMC  Etiology of Severe Acute Respiratory Infection...   \n",
       "14601  embv1el9      PMC  Accuracy of a battery-powered portable capnome...   \n",
       "14602  8dcgzyst      PMC  The influence of diabetes on postoperative com...   \n",
       "14603  sgtvecj8      PMC  In vivo delivery of a multiepitope peptide and...   \n",
       "14604  3vtq6d8f      PMC  Relationship Between Compassion Fatigue in Nur...   \n",
       "\n",
       "      license                                           abstract publish_time  \\\n",
       "14597   no-cc  in april 2017, surveillance detected a surge i...   2021-01-27   \n",
       "14601   no-cc  carbon dioxide measurement is useful for confi...   2021-01-01   \n",
       "14602   no-cc  background: diabetes mellitus has been commonl...   2021-01-01   \n",
       "14603   no-cc  objectives: a potent hiv vaccine should overco...   2021-01-01   \n",
       "14604   no-cc  this research was conducted in order to examin...   2021-01-02   \n",
       "\n",
       "                                                 authors              journal  \\\n",
       "14597  Rahaman, Md R.; Alroy, Karen A.; Van Beneden, ...     Emerg Infect Dis   \n",
       "14601                     Hirakawa, Eiji; Ibara, Satoshi  J Clin Monit Comput   \n",
       "14602  Tan, D. J. H.; Yaow, C. Y. L.; Mok, H. T.; Ng,...     Tech Coloproctol   \n",
       "14603   Davoodi, Saba; Bolhassani, Azam; Namazi, Fatemeh      Biotechnol Lett   \n",
       "14604       Aslan, Hakime; Erci, Behice; Pekince, Hatice       J Relig Health   \n",
       "\n",
       "                                          pdf_json_files  \\\n",
       "14597  document_parses/pdf_json/3798d3bcb09df78296832...   \n",
       "14601  document_parses/pdf_json/2b4690b639d799f33bbf7...   \n",
       "14602  document_parses/pdf_json/83a841d74d4f6bf8bffb8...   \n",
       "14603  document_parses/pdf_json/7dd8097ba18590caa4310...   \n",
       "14604  document_parses/pdf_json/9bd542f156f108ef170a0...   \n",
       "\n",
       "                                         abstract_tokens  \n",
       "14597  [april, 2017, surveillance, detected, surge, s...  \n",
       "14601  [carbon, dioxide, measurement, useful, confirm...  \n",
       "14602  [background, diabetes, mellitus, commonly, ass...  \n",
       "14603  [objectives, potent, hiv, vaccine, overcome, l...  \n",
       "14604  [research, conducted, order, examine, correlat...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# tokenize abstract for data search\n",
    "\n",
    "data1['abstract'] = data1['abstract'].apply(lambda x: x.strip().lower())\n",
    "stopwords = set(stopwords.words('english'))\n",
    "data1['abstract_tokens'] = data1['abstract'].apply(lambda x: word_tokenize(x))\n",
    "\n",
    "## Remove Stopwords\n",
    "data1['abstract_tokens'] = data1['abstract_tokens'].apply(lambda x: [w for w in x if w not in stopwords])\n",
    "\n",
    "## Remove words with single characters\n",
    "data1['abstract_tokens'] = data1['abstract_tokens'].apply(lambda x: [w for w in x if len(w)>1])\n",
    "data1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cord_uid</th>\n",
       "      <th>source_x</th>\n",
       "      <th>title</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>authors</th>\n",
       "      <th>journal</th>\n",
       "      <th>pdf_json_files</th>\n",
       "      <th>abstract_tokens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>aoiambb8</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Etiology of Severe Acute Respiratory Infection...</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>in april 2017, surveillance detected a surge i...</td>\n",
       "      <td>2021-01-27</td>\n",
       "      <td>Rahaman, Md R.; Alroy, Karen A.; Van Beneden, ...</td>\n",
       "      <td>Emerg Infect Dis</td>\n",
       "      <td>document_parses/pdf_json/3798d3bcb09df78296832...</td>\n",
       "      <td>[april, 2017, surveillance, detected, surge, s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>embv1el9</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Accuracy of a battery-powered portable capnome...</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>carbon dioxide measurement is useful for confi...</td>\n",
       "      <td>2021-01-01</td>\n",
       "      <td>Hirakawa, Eiji; Ibara, Satoshi</td>\n",
       "      <td>J Clin Monit Comput</td>\n",
       "      <td>document_parses/pdf_json/2b4690b639d799f33bbf7...</td>\n",
       "      <td>[carbon, dioxide, measurement, useful, confirm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8dcgzyst</td>\n",
       "      <td>PMC</td>\n",
       "      <td>The influence of diabetes on postoperative com...</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>background: diabetes mellitus has been commonl...</td>\n",
       "      <td>2021-01-01</td>\n",
       "      <td>Tan, D. J. H.; Yaow, C. Y. L.; Mok, H. T.; Ng,...</td>\n",
       "      <td>Tech Coloproctol</td>\n",
       "      <td>document_parses/pdf_json/83a841d74d4f6bf8bffb8...</td>\n",
       "      <td>[background, diabetes, mellitus, commonly, ass...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sgtvecj8</td>\n",
       "      <td>PMC</td>\n",
       "      <td>In vivo delivery of a multiepitope peptide and...</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>objectives: a potent hiv vaccine should overco...</td>\n",
       "      <td>2021-01-01</td>\n",
       "      <td>Davoodi, Saba; Bolhassani, Azam; Namazi, Fatemeh</td>\n",
       "      <td>Biotechnol Lett</td>\n",
       "      <td>document_parses/pdf_json/7dd8097ba18590caa4310...</td>\n",
       "      <td>[objectives, potent, hiv, vaccine, overcome, l...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3vtq6d8f</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Relationship Between Compassion Fatigue in Nur...</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>this research was conducted in order to examin...</td>\n",
       "      <td>2021-01-02</td>\n",
       "      <td>Aslan, Hakime; Erci, Behice; Pekince, Hatice</td>\n",
       "      <td>J Relig Health</td>\n",
       "      <td>document_parses/pdf_json/9bd542f156f108ef170a0...</td>\n",
       "      <td>[research, conducted, order, examine, correlat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cord_uid source_x                                              title  \\\n",
       "0  aoiambb8      PMC  Etiology of Severe Acute Respiratory Infection...   \n",
       "1  embv1el9      PMC  Accuracy of a battery-powered portable capnome...   \n",
       "2  8dcgzyst      PMC  The influence of diabetes on postoperative com...   \n",
       "3  sgtvecj8      PMC  In vivo delivery of a multiepitope peptide and...   \n",
       "4  3vtq6d8f      PMC  Relationship Between Compassion Fatigue in Nur...   \n",
       "\n",
       "  license                                           abstract publish_time  \\\n",
       "0   no-cc  in april 2017, surveillance detected a surge i...   2021-01-27   \n",
       "1   no-cc  carbon dioxide measurement is useful for confi...   2021-01-01   \n",
       "2   no-cc  background: diabetes mellitus has been commonl...   2021-01-01   \n",
       "3   no-cc  objectives: a potent hiv vaccine should overco...   2021-01-01   \n",
       "4   no-cc  this research was conducted in order to examin...   2021-01-02   \n",
       "\n",
       "                                             authors              journal  \\\n",
       "0  Rahaman, Md R.; Alroy, Karen A.; Van Beneden, ...     Emerg Infect Dis   \n",
       "1                     Hirakawa, Eiji; Ibara, Satoshi  J Clin Monit Comput   \n",
       "2  Tan, D. J. H.; Yaow, C. Y. L.; Mok, H. T.; Ng,...     Tech Coloproctol   \n",
       "3   Davoodi, Saba; Bolhassani, Azam; Namazi, Fatemeh      Biotechnol Lett   \n",
       "4       Aslan, Hakime; Erci, Behice; Pekince, Hatice       J Relig Health   \n",
       "\n",
       "                                      pdf_json_files  \\\n",
       "0  document_parses/pdf_json/3798d3bcb09df78296832...   \n",
       "1  document_parses/pdf_json/2b4690b639d799f33bbf7...   \n",
       "2  document_parses/pdf_json/83a841d74d4f6bf8bffb8...   \n",
       "3  document_parses/pdf_json/7dd8097ba18590caa4310...   \n",
       "4  document_parses/pdf_json/9bd542f156f108ef170a0...   \n",
       "\n",
       "                                     abstract_tokens  \n",
       "0  [april, 2017, surveillance, detected, surge, s...  \n",
       "1  [carbon, dioxide, measurement, useful, confirm...  \n",
       "2  [background, diabetes, mellitus, commonly, ass...  \n",
       "3  [objectives, potent, hiv, vaccine, overcome, l...  \n",
       "4  [research, conducted, order, examine, correlat...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data1.reset_index(drop=True,inplace=True)\n",
    "data1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 312151 unique words in inverted index\n",
      "Found 42340 articles related to covid\n",
      "42340\n"
     ]
    }
   ],
   "source": [
    "# Filter topic\n",
    "\n",
    "inverted_idx = {}\n",
    "for idx,sentence in enumerate(list(data1['abstract_tokens'])):\n",
    "    for token in set(sentence):\n",
    "        if token not in inverted_idx:\n",
    "            inverted_idx[token] = [idx]\n",
    "        else:\n",
    "            inverted_idx[token].append(idx)\n",
    "\n",
    "print(f'There are {len(inverted_idx)} unique words in inverted index')\n",
    "\n",
    "keywords_vaccine = ['covid','covid-19']\n",
    "vaccine_idx = []\n",
    "for word in keywords_vaccine:\n",
    "    try:\n",
    "        vaccine_idx += inverted_idx[word]\n",
    "    except:\n",
    "        pass\n",
    "vaccine_idx = list(set(vaccine_idx))\n",
    "print(f'Found {len(vaccine_idx)} articles related to covid')\n",
    "data2 = data1.loc[vaccine_idx]\n",
    "print(len(data2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 177597 unique words in inverted index\n",
      "Found 3364 articles related to vaccine\n",
      "3364\n"
     ]
    }
   ],
   "source": [
    "data2.reset_index(drop=True,inplace=True)\n",
    "\n",
    "inverted_idx = {}\n",
    "for idx,sentence in enumerate(list(data2['abstract_tokens'])):\n",
    "    for token in set(sentence):\n",
    "        if token not in inverted_idx:\n",
    "            inverted_idx[token] = [idx]\n",
    "        else:\n",
    "            inverted_idx[token].append(idx)\n",
    "\n",
    "print(f'There are {len(inverted_idx)} unique words in inverted index')\n",
    "\n",
    "keywords_vaccine = ['vaccine']\n",
    "vaccine_idx = []\n",
    "for word in keywords_vaccine:\n",
    "    try:\n",
    "        vaccine_idx += inverted_idx[word]\n",
    "    except:\n",
    "        pass\n",
    "vaccine_idx = list(set(vaccine_idx))\n",
    "print(f'Found {len(vaccine_idx)} articles related to vaccine')\n",
    "data3 = data2.loc[vaccine_idx]\n",
    "print(len(data3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 32456 unique words in inverted index\n",
      "Found 296 articles related to adverse\n",
      "296\n"
     ]
    }
   ],
   "source": [
    "data3.reset_index(drop=True,inplace=True)\n",
    "\n",
    "inverted_idx = {}\n",
    "for idx,sentence in enumerate(list(data3['abstract_tokens'])):\n",
    "    for token in set(sentence):\n",
    "        if token not in inverted_idx:\n",
    "            inverted_idx[token] = [idx]\n",
    "        else:\n",
    "            inverted_idx[token].append(idx)\n",
    "\n",
    "print(f'There are {len(inverted_idx)} unique words in inverted index')\n",
    "\n",
    "#keywords_vaccine = ['saftey','safe','surveillance','reactogenicity','clinial','trial','trials']\n",
    "#keywords_vaccine = ['saftey','safe','surveillance','reactogenicity']\n",
    "keywords_vaccine = ['adverse']\n",
    "vaccine_idx = []\n",
    "for word in keywords_vaccine:\n",
    "    try:\n",
    "        vaccine_idx += inverted_idx[word]\n",
    "    except:\n",
    "        pass\n",
    "vaccine_idx = list(set(vaccine_idx))\n",
    "print(f'Found {len(vaccine_idx)} articles related to adverse')\n",
    "data4 = data3.loc[vaccine_idx]\n",
    "print(len(data4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data4.to_pickle(r'C:\\Users\\Gerard\\Desktop\\ISS\\Pattern Recognition Systems\\4 Practice module\\Cord19\\archive/metadata_filter.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pre-processing corpus for SBERT & TFIDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data4 = pd.read_pickle(r'C:\\Users\\Gerard\\Desktop\\ISS\\Pattern Recognition Systems\\4 Practice module\\Cord19\\archive/metadata_filter.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import trained SBERT model\n",
    "\n",
    "tokenizer = AutoTokenizer.from_pretrained(\"gerardozq/biobert_v1.1_pubmed-finetuned-squad\")\n",
    "model = AutoModelForQuestionAnswering.from_pretrained(\"gerardozq/biobert_v1.1_pubmed-finetuned-squad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6998\n",
      "6998\n"
     ]
    }
   ],
   "source": [
    "# Process corpus for SBERT\n",
    "\n",
    "\n",
    "train_sent1 = []\n",
    "train_sent2 = []\n",
    "train_label = []\n",
    "citecorp = []\n",
    "paracorp = []\n",
    "para_doc = {}\n",
    "para_idx = 0\n",
    "\n",
    "j=0\n",
    "for i in range(0,data4.shape[0]):\n",
    "    #print('archive/'+data4['pdf_json_files'].iloc[i])\n",
    "    filename = 'archive/'+data4['pdf_json_files'].iloc[i]\n",
    "    filename = filename.split(\"; \")\n",
    "    #print(filename[0])\n",
    "    with open(filename[0]) as f_json:\n",
    "        full_text_dict = json.load(f_json)\n",
    "        for paragraph_dict in full_text_dict['body_text']:\n",
    "            if len(paragraph_dict['text'])>300:\n",
    "                ref_text = []\n",
    "                paragraph_text = paragraph_dict['text']\n",
    "                for ref in paragraph_dict['cite_spans']:\n",
    "                    ref_id = ref['ref_id']\n",
    "                    try:\n",
    "                        ref_title = full_text_dict['bib_entries'][ref_id]['title']\n",
    "                    except:\n",
    "                        ref_title = full_text_dict['metadata']['title']\n",
    "                    ref_text.append(ref_title)\n",
    "                for trainref in ref_text:\n",
    "                    train_sent1.append(paragraph_text)\n",
    "                    train_sent2.append(trainref)\n",
    "                    train_label.append(1.0)\n",
    "                    if trainref not in citecorp:\n",
    "                        citecorp.append(trainref)\n",
    "                paracorp.append(paragraph_text)\n",
    "                para_doc[para_idx] = {}\n",
    "                para_doc[para_idx]['title'] = data4['title'].iloc[i]\n",
    "                para_doc[para_idx]['text'] = paragraph_text\n",
    "                para_doc[para_idx]['authors'] = data4['authors'].iloc[i]\n",
    "                para_doc[para_idx]['publish_time'] = data4['publish_time'].iloc[i]\n",
    "                para_doc[para_idx]['journal'] = data4['journal'].iloc[i]\n",
    "                para_idx = para_idx + 1\n",
    "                \n",
    "\n",
    "print(len(para_doc))\n",
    "print(len(paracorp))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6998\n",
      "6998\n"
     ]
    }
   ],
   "source": [
    "# Process corpus for TFIDF\n",
    "\n",
    "doc_list = []\n",
    "doc_list_word = []\n",
    "\n",
    "for i in range(0,data4.shape[0]):\n",
    "    doc_text = []\n",
    "    doc_text_word = []\n",
    "    #print('archive/'+data4['pdf_json_files'].iloc[i])\n",
    "    filename = 'archive/'+data4['pdf_json_files'].iloc[i]\n",
    "    filename = filename.split(\"; \")\n",
    "    #print(filename[0])\n",
    "    with open(filename[0]) as f_json:\n",
    "        full_text_dict = json.load(f_json)\n",
    "        for paragraph_dict in full_text_dict['body_text']:\n",
    "            if len(paragraph_dict['text'])>300:\n",
    "                paragraph_text = paragraph_dict['text']\n",
    "                text_token = tokenizer(paragraph_text[:512])\n",
    "                doc_list_word.append(paragraph_text)\n",
    "                doc_list.append(text_token['input_ids'])\n",
    "print(len(doc_list))\n",
    "print(len(doc_list_word))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export processed corpus if needed\n",
    "import pickle\n",
    "tfidf_doc = {}\n",
    "tfidf_doc['doc_list'] = doc_list\n",
    "tfidf_doc['doc_list_word'] = doc_list_word\n",
    "\n",
    "filename = 'tfidf_doc'\n",
    "outfile = open(filename,'wb')\n",
    "pickle.dump(tfidf_doc,outfile)\n",
    "outfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export processed corpus if needed\n",
    "import pickle\n",
    "sbert_doc = {}\n",
    "sbert_doc['paracorp'] = paracorp\n",
    "sbert_doc['para_doc'] = para_doc\n",
    "\n",
    "filename = 'sbert_doc'\n",
    "outfile = open(filename,'wb')\n",
    "pickle.dump(sbert_doc,outfile)\n",
    "outfile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create SBERT embeddings for export to system"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "embedder = SentenceTransformer('msmarco-distilbert-base-v4')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Encode\n",
    "corpus = paracorp\n",
    "\n",
    "corpus_embeddings = embedder.encode(corpus, convert_to_tensor=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export SBERT embeddings\n",
    "import pickle\n",
    "\n",
    "filename = 'emb_corpus300'\n",
    "outfile = open(filename,'wb')\n",
    "pickle.dump(corpus_embeddings,outfile)\n",
    "outfile.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "filename = 'emb_corpus300'\n",
    "infile = open(filename,'rb')\n",
    "corpus_embeddings1 = pickle.load(infile)\n",
    "infile.close()\n",
    "\n",
    "corpus = paracorp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optional: Export all models "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "models_litmining = {}\n",
    "models_litmining['tokenizer'] = tokenizer\n",
    "models_litmining['model'] = model\n",
    "models_litmining['embedder'] = embedder\n",
    "\n",
    "filename = 'models_litmining'\n",
    "outfile = open(filename,'wb')\n",
    "pickle.dump(models_litmining,outfile)\n",
    "outfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "filename = 'models_litmining'\n",
    "infile = open(filename,'rb')\n",
    "models_litmining1 = pickle.load(infile)\n",
    "infile.close()\n",
    "\n",
    "tokenizer = models_litmining1['tokenizer']\n",
    "model = models_litmining1['model']\n",
    "embedder = models_litmining1['embedder']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dummy_fun(doc):\n",
    "    return doc\n",
    "\n",
    "tfidf = TfidfVectorizer(\n",
    "    analyzer='word',\n",
    "    tokenizer=dummy_fun,\n",
    "    preprocessor=dummy_fun,\n",
    "    token_pattern=None)  \n",
    "\n",
    "tfidf_matrix = tfidf.fit_transform(doc_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cord19mining(symptomInput = ['pain']):\n",
    "\n",
    "    # symptom processing\n",
    "    symptomlist = []\n",
    "    rejectlist = ['covid-19', 'blood test','computerised tomogram','sars-cov-2 test positive',\n",
    "                  'cerebrovascular accident','electrocardiogram','echocardiogram','troponin increased',\n",
    "                 'arthralgia','hyperhidrosis','paraesthesia','hypoaesthesia','feeling abnormal']\n",
    "    for symptom in symptomInput:\n",
    "        if symptom == 'dyspnoea':\n",
    "            symptom1 = 'dyspnea'\n",
    "        elif symptom == 'pyrexia':\n",
    "            symptom1 = 'fever'\n",
    "        elif symptom == 'injection site erythema':\n",
    "            symptom1 = 'erythema'\n",
    "        elif symptom == 'myalgia':\n",
    "            symptom1 = 'myalgias'\n",
    "        elif symptom == 'lymphadenopathy':\n",
    "            symptom1 = 'lymph nodes'\n",
    "        elif symptom in rejectlist:\n",
    "            symptom1 = 'pain'\n",
    "        else:\n",
    "            symptom1 = symptom\n",
    "        \n",
    "        if symptom1 not in symptomlist:\n",
    "            symptomlist.append(symptom1)\n",
    "        \n",
    "    \n",
    "    output = {}\n",
    "    i = 1\n",
    "    \n",
    "    resultlist_value = torch.empty((0))\n",
    "    resultlist_indices = torch.empty((0))\n",
    "    query_symp = []\n",
    "\n",
    "    for symp in symptomlist:\n",
    "        # Auto-query generation\n",
    "        query = 'is ' + str(symp) + ' caused by vaccine a severe adverse effect'\n",
    "        #queries = ['is fever caused by vaccine a severe adverse effect']\n",
    "        \n",
    "        # tfidf score\n",
    "        query_token = tokenizer(query)\n",
    "        query_vec = tfidf.transform([query_token['input_ids'][1:len(query_token['input_ids'])-1]])\n",
    "        cosine_sim = cosine_similarity(tfidf_matrix, query_vec)\n",
    "        tfidf_score = torch.FloatTensor(np.transpose(cosine_sim)[0])\n",
    "\n",
    "        # sbert score\n",
    "        top_k = min(5, len(corpus))\n",
    "\n",
    "        query_embedding = embedder.encode(query, convert_to_tensor=True)\n",
    "        \n",
    "        k_coeff = 0.7\n",
    "        \n",
    "        # Linear combination of tfidf and sbert scores to find the highest 5 scores\n",
    "        cos_scores = util.pytorch_cos_sim(query_embedding, corpus_embeddings1)[0]\n",
    "        combined_scores = k_coeff*cos_scores + (1-k_coeff)*tfidf_score\n",
    "        top_results = torch.topk(combined_scores, k=top_k)\n",
    "        resultlist_value = torch.cat((resultlist_value,top_results[0]))\n",
    "        resultlist_indices = torch.cat((resultlist_indices,top_results[1]))\n",
    "        query_symp.extend([symp for i in range(5)])\n",
    "    \n",
    "    top_results_list = torch.topk(resultlist_value, k=top_k)\n",
    "    symp_idx = top_results_list[1].tolist()\n",
    "    query_symp = [query_symp[i] for i in symp_idx]\n",
    "    qi = 0\n",
    "\n",
    "    \n",
    "    for idx in resultlist_indices[top_results_list[1]].int():\n",
    "        # QA search\n",
    "        question = 'is ' + query_symp[qi] + ' caused by vaccine a severe adverse effect'\n",
    "        text = corpus[idx]\n",
    "        inputs = tokenizer(question, text, add_special_tokens=True, return_tensors=\"pt\")\n",
    "        input_ids = inputs[\"input_ids\"].tolist()[0]\n",
    "        outputs = model(**inputs)\n",
    "        answer_start_scores = outputs.start_logits\n",
    "        answer_end_scores = outputs.end_logits\n",
    "        # Get the most likely beginning of answer with the argmax of the score\n",
    "        answer_start = torch.argmax(answer_start_scores)\n",
    "        # Get the most likely end of answer with the argmax of the score\n",
    "        answer_end = torch.argmax(answer_end_scores) + 1\n",
    "        answer = tokenizer.convert_tokens_to_string(tokenizer.convert_ids_to_tokens(input_ids[answer_start:answer_end]))\n",
    "        \n",
    "        # Sentiment analyzer\n",
    "        analyzer = SentimentIntensityAnalyzer()\n",
    "        sentence = answer \n",
    "        vs = analyzer.polarity_scores(sentence)\n",
    "        \n",
    "        output[i] = {}\n",
    "        output[i]['text'] =  para_doc[idx.item()]['text']\n",
    "        output[i]['title'] =  para_doc[idx.item()]['title']\n",
    "        output[i]['authors'] =  para_doc[idx.item()]['authors']\n",
    "        output[i]['publish_time'] = para_doc[idx.item()]['publish_time'].strftime('%m/%d/%Y')     \n",
    "        output[i]['journal'] =  para_doc[idx.item()]['journal']\n",
    "        output[i]['sentiment'] =  vs\n",
    "        i = i + 1\n",
    "        qi = qi + 1\n",
    "        \n",
    "    return output\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: {'text': 'Vaccines activate the immune system, which will commonly result in minor side effects, including mild fever and local inflammatory reactions at the site of the injection. This may include redness, swelling, pain, and warmth at the injection sites [1] . These reactions are not a contraindication to receiving the same vaccine in the future, as they do not pose a risk for future allergic reactions to the vaccine. Non-allergic reactions to vaccines also include anxiety-related adverse events that can mimic allergic reactions, and may include breath-holding, hyperventilation, and vasovagal syncope (fainting) (see Table 1 in the Canadian Immunization Guide: Anaphylaxis and other Acute Reactions following Vaccination) [2] .',\n",
       "  'title': 'COVID-19 vaccine testing & administration guidance for allergists/immunologists from the Canadian Society of Allergy and Clinical Immunology (CSACI)',\n",
       "  'authors': 'Vander Leek, Timothy K.; Chan, Edmond S.; Connors, Lori; Derfalvi, Beata; Ellis, Anne K.; Upton, Julia E. M.; Abrams, Elissa M.',\n",
       "  'publish_time': '03/15/2021',\n",
       "  'journal': 'Allergy Asthma Clin Immunol',\n",
       "  'sentiment': {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}},\n",
       " 2: {'text': 'In this phase 1/2 clinical trial among healthy adults, the SARS-CoV-2 inactivated vaccine was safe and well-tolerated in all age and dose groups, and 15¢1% of participants experienced adverse reactions. The most common injection-site adverse reaction was injection-site pain, and the most common systemic adverse reaction was fever, which were mostly mild-to-moderate and self-limiting. No serious adverse reactions related to the vaccine occurred. The inactivated vaccine could effectively elicit humoral responses, and the GMT of neutralizing antibody titres on days 28 and 90 after the third injection ranged from 149 to 238 and 87 to 129, respectively among participants receiving three doses of vaccines. In general, the antibody response was weaker in 5 mg days 0 and 14 group and 10 mg day 0 group. Given that large-scale production of the inactivated vaccine would be achievable and the vaccine could be stored at 2À8°C, the inactivated vaccine could be an optimal option for many developing countries and regions where cold storage and distribution capacity was relatively poor.',\n",
       "  'title': 'Safety and immunogenicity of an inactivated SARS-CoV-2 vaccine in healthy adults aged 18 years or older: A randomized, double-blind, placebo-controlled, phase 1/2 trial',\n",
       "  'authors': 'Guo, Wanshen; Duan, Kai; Zhang, Yuntao; Yuan, Zhiming; Zhang, Yan-Bo; Wang, Zejun; Zhao, Dongyang; Zhang, Huajun; Xie, Zhiqiang; Li, Xinguo; Peng, Cheng; Zhang, Wei; Yang, Yunkai; Chen, Wei; Gao, Xiaoxiao; You, Wangyang; Wang, Xue-Wei; Shi, Zhengli; Wang, Yanxia; Yang, Xu-Qin; Zhang, Lianghao; Huang, Lili; Wang, Qian; Lu, Jia; Yang, Yong-Li; Guo, Jing; Zhou, Wei; Wan, Xin; Wu, Cong; Wang, Wenhui; Du, Jianhui; Nian, Xuanxuan; Li, Xing-Hang; Huang, Shihe; Shen, Shuo; Xia, Shengli; Pan, An; Yang, Xiaoming',\n",
       "  'publish_time': '07/07/2021',\n",
       "  'journal': 'EClinicalMedicine',\n",
       "  'sentiment': {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}},\n",
       " 3: {'text': \"Overall, the most frequent adverse effects reported after the administration of the three vaccines considered in this review consisted of local reactions at the injection site (sore arm and erythema) followed by non-specific systemic effects (myalgia, chills, fatigue, headache, and fever). These adverse reactions are considered as the result of the immune system's reaction that occurs soon after vaccination (also referred to as reactogenicity) and resolves shortly [35] . Given the scale-up of mass vaccination campaigns across the world, severe adverse events may occur and will likely generate concerns among patients and requests for evaluation.\",\n",
       "  'title': 'Safety of COVID-19 vaccines administered in the EU: Should we be concerned?',\n",
       "  'authors': 'Hernández, Antonio F.; Calina, Daniela; Poulas, Konstantinos; Docea, Anca Oana; Tsatsakis, Aristidis M.',\n",
       "  'publish_time': '04/20/2021',\n",
       "  'journal': 'Toxicol Rep',\n",
       "  'sentiment': {'neg': 0.267, 'neu': 0.733, 'pos': 0.0, 'compound': -0.6249}},\n",
       " 4: {'text': 'Severe allergic reactions due to anaphylaxis represent a very rare side effect associated with most vaccines, including mRNA vaccines. For this reason, it has been recommended that subjects with a previous history of severe allergic reactions to any vaccine ingredient should not receive it [36] . This safety concern is currently followed-up via routine pharmacovigilance.',\n",
       "  'title': 'Safety of COVID-19 vaccines administered in the EU: Should we be concerned?',\n",
       "  'authors': 'Hernández, Antonio F.; Calina, Daniela; Poulas, Konstantinos; Docea, Anca Oana; Tsatsakis, Aristidis M.',\n",
       "  'publish_time': '04/20/2021',\n",
       "  'journal': 'Toxicol Rep',\n",
       "  'sentiment': {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}},\n",
       " 5: {'text': 'In the population analysed, only one serious adverse event (0.05%) was reported in an approximately 30-year-old female subject, who was immediately hospitalised after the second dose of vaccine due to symptoms similar to anaphylaxis (skin rash and glottal oedema). This serious adverse reaction was completely resolved within 24 hours.',\n",
       "  'title': 'Comparison of medium-term adverse reactions induced by the first and second dose of mRNA BNT162b2 (Comirnaty, Pfizer-BioNTech) vaccine: a post-marketing Italian study conducted between 1 January and 28 February 2021',\n",
       "  'authors': 'Ossato, Andrea; Tessari, Roberto; Trabucchi, Carlotta; Zuppini, Teresa; Realdon, Nicola; Marchesini, Francesca',\n",
       "  'publish_time': '07/26/2021',\n",
       "  'journal': 'Eur J Hosp Pharm',\n",
       "  'sentiment': {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}}}"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cord19mining(symptomInput = ['pyrexia','injection site warmth'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output results are provided to system UI"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
